docs/enhancements/auto-return-agent-to-infra-env.md

```diff
@@ -0,0 +1,247 @@
+            |                                               |
+            V                                               |
+  ------------------------                                  |
+  | HostStatusReclaiming |                                  |
+  ------------------------                                  |
+                |                                           |
+       Agent posts step reply                               |
+                |                                           |
+       ---------+----------------------------------         |
+       |                                          |         |
+    Success                                    Failure      |
+       |                                          |         |
+       V                                          |         |
+-------------------------------------             |         |
+| HostStatusReclaimingPendingReboot |------       |         |
+-------------------------------------     |       |         |
+                |                         |       |         |
+  ------------------------------       Timeout    |         |
+  | TransitionTypeRegisterHost |          |       |         |
+  ------------------------------          |       |         |
+                |                         |       |         |
+                V                         V       V         V
+--------------------------------  ----------------------------------------
+| HostStatusDiscoveringUnbound |  | HostStatusUnbindingPendingUserAction |
+--------------------------------  ----------------------------------------
+              ^                                       |
+              |       ------------------------------  |
+              --------| TransitionTypeRegisterHost |---
+                      ------------------------------
+```
+
+#### Assumptions
+
+* This approach only works if the host is currently part of a working cluster
+  - This cluster needs to have been installed by assisted service so that
+    it has access to the kubeconfig
+* This reclaim procedure will only be triggered if assisted service is running
+  the kubernetes API controllers
+  - Assisted service needs to ensure the host isn't being managed by a bare
+    metal host (only possible with the kubernetes API)
+  - When using the REST API there is no way for a user to indicate if a host
+    should be unbound normally or if it should be reclaimed
+* There must be enough free space on the `/boot` partition
+  - Space is required to store the initrd and kernel (~90MB)
+  - The agent can check for this
+* A node in a spoke cluster can be identified by the hostname in the agent
+  - In a non-CAPI scenario there are no other reliable means to identify which
+    node maps to which agent
+  - With CAPI a string of references exist that could be used to identify the
+    node, but the hostname should be reliable in this case as well
```

Comment:
We could also have some tiered logic where we first check for the CAPI machine resources, then check for the system uid then fall back to hostname, but hostname should just work all the time, right? Does anyone know of a reason it wouldn't?

@mhrivnak @avishayt @tsorya @eranco74 ?